lib/task_server.mli

```diff
@@ -0,0 +1,106 @@
+(* Task_server: Helper for interfaces of a particular shape
+   *
+   * This module is a helper module for interfaces that implement a Task interface
+   * with which asynchronous calls are implemented. In particular, it handles
+   * subtasks, cancellation, reporting of backtraces and a way of injecting
+   * failures at particular points known as 'cancellation triggers'
+*)
+
+type stringpair = string * string
+
+module type INTERFACE =
+sig
+  val service_name : string
+  exception Does_not_exist of stringpair
+  exception Cancelled of string
+  module Task :
+  sig
+    type id = string
+    type async_result
+    val rpc_of_async_result : async_result -> Rpc.t
+    val async_result_of_rpc : Rpc.t -> async_result
+    type completion_t = {
+      duration : float;
+      result : async_result option;
+    }
+    type state =
+        Pending of float
+      | Completed of completion_t
+      | Failed of Rpc.t
+    type t = {
+      id: id;
+      dbg: string;
+      ctime: float;
+      state: state;
+      subtasks: (string * state) list;
+      debug_info: (string * string) list;
+      backtrace: string;
+    }
+  end
+  module Exception : sig type exnty val rpc_of_exnty : exnty -> Rpc.t end
+  val exnty_of_exn : exn -> Exception.exnty
+  val exn_of_exnty : Exception.exnty -> exn
+  exception Internal_error of string
+end
+
+module Task :
+  functor (Interface : INTERFACE) ->
+  sig
+
+    (* An [id] is a marshallable reference to a task in a [tasks] type *)
+    type id = string
+
+    (* A [task_handle] is required for all operations on tasks *)
+    type task_handle
+
+    (* A [tasks] record contains a list of tasks *)
+    type tasks
```

Comment:
Yes perhaps, but I didn't want to break the interface even more... see https://github.com/xapi-project/xenopsd/pull/320/files for the current fallout of just these changes.